compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerToUKernels.cpp

```diff
@@ -152,6 +152,176 @@ static FailureOr<IREE::Codegen::UKernelOpInterface> matchDAGForUKernel(
+    return rewriter.notifyMatchFailure(op, "unsupported outer_dims_perm");
+  }
+
+  Location loc = op.getLoc();
+  Value paddingVal = op.getPaddingValue();
+  if (!paddingVal) {
+    paddingVal = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(inType.getElementType()),
+        inType.getElementType());
+  }
+
+  Value in_size0 = rewriter.create<tensor::DimOp>(loc, in, 0);
+  Value in_size1 = rewriter.create<tensor::DimOp>(loc, in, 1);
+  Value out_size0 = rewriter.create<tensor::DimOp>(loc, out, 0);
+  Value out_size1 = rewriter.create<tensor::DimOp>(loc, out, 1);
+  Value out_size2 = rewriter.create<tensor::DimOp>(loc, out, 2);
+  Value out_size3 = rewriter.create<tensor::DimOp>(loc, out, 3);
+  Value flagsVal = rewriter.create<arith::ConstantOp>(
+      loc, rewriter.getI32IntegerAttr(flags));
+  auto genericMicroKernelOp = rewriter.create<IREE::Codegen::UKernelGenericOp>(
+      loc, outType, fnName, in, out,
+      ValueRange{in_size0, in_size1, out_size0, out_size1, out_size2, out_size3,
+                 paddingVal, flagsVal},
+      /*strided_outer_dims=*/rewriter.getIndexAttr(1));
+  return cast<IREE::Codegen::UKernelOpInterface>(
+      genericMicroKernelOp.getOperation());
+}
+
+static FailureOr<IREE::Codegen::UKernelOpInterface> matchDAGForUKernel(
+    RewriterBase &rewriter, tensor::UnPackOp op) {
+  if (!op.hasTensorSemantics()) {
+    return rewriter.notifyMatchFailure(
+        op, "operations needs to have tensor semantics");
+  }
+  Value in = op.getSource();
+  Value out = op.getDest();
+  auto inType = in.getType().cast<ShapedType>();
+  auto outType = out.getType().cast<ShapedType>();
+  Type inElemType = inType.getElementType();
+  Type outElemType = outType.getElementType();
+  std::string fnName;
+  if (inElemType.isSignlessInteger(32) && outElemType.isSignlessInteger(32)) {
+    fnName = "vmvx.unpack.i32i32";
+  } else if (inElemType.isF32() && outElemType.isF32()) {
+    fnName = "vmvx.unpack.f32f32";
+  } else {
+    return rewriter.notifyMatchFailure(
+        op, "unsupported combination of element types");
+  }
+
+  if (inType.getRank() != 4) {
+    return rewriter.notifyMatchFailure(op, "expected input to be 4D");
+  }
+
+  if (outType.getRank() != 2) {
+    return rewriter.notifyMatchFailure(op, "expected output to be 2D");
+  }
+
+  int64_t innerDimsPos[2] = {0, 1};
+  ArrayRef<int64_t> innerDimsPosArr = op.getInnerDimsPos();
+  if (!innerDimsPosArr.empty()) {
+    innerDimsPos[0] = innerDimsPosArr[0];
+    innerDimsPos[1] = innerDimsPosArr[1];
+  }
+
+  int64_t outerDimsPerm[2] = {0, 1};
+  ArrayRef<int64_t> outerDimsPosArr = op.getOuterDimsPerm();
+  if (!outerDimsPosArr.empty()) {
+    outerDimsPerm[0] = outerDimsPosArr[0];
+    outerDimsPerm[1] = outerDimsPosArr[1];
+  }
+
+  int flags = 0;
+
+  if (innerDimsPos[0] == 0 && innerDimsPos[1] == 1) {
+    // nothing to do
+  } else if (innerDimsPos[0] == 1 && innerDimsPos[1] == 0) {
+    flags |= IREE_UK_FLAG_PACK_TRANSPOSE_INNER;
```

Comment:
should this be `IREE_UK_FLAG_UNPACK_TRANSPOSE_INNER`?

Reply:
Wow, great catch. This is what typed enums are for, but here we have to use raw integer literals as this is shared with C and even potentially with assembly (see `exported_bits.h`).